server/src/main/java/org/apache/druid/server/coordinator/duty/KillUnusedSegments.java

```diff
@@ -204,7 +230,10 @@ private void killUnusedSegments(
         );
         ++submittedTasks;
         datasourceToLastKillIntervalEnd.put(dataSource, intervalToKill.getEnd());
-        remainingDatasourcesToKill.remove(dataSource);
+
+        if (remainingDatasourcesToKill.isEmpty() || submittedTasks >= availableKillTaskSlots) {
+          break;
+        }
```

Comment:
Would it be better to have this check right at the beginning of the loop?
Currently there are multiple breaks and continue which makes the code a little difficult to follow.

Reply:
Yeah, we can. However, it will slightly change the semantics of the round-robin iteration, i.e., the loop will terminate _after_ seeing a new element without doing anything with it.

I've added brief line comments next to these conditions for now.